lib/logstash/filters/mutate.rb

```diff
@@ -338,85 +334,91 @@ def gsub_dynamic_fields(event, original, needle, replacement)
     end
   end
 
-  private
   def uppercase(event)
     @uppercase.each do |field|
-      if event[field].is_a?(Array)
-        event[field].each { |v| v.upcase! }
-      elsif event[field].is_a?(String)
-        event[field].upcase!
-      else
-        @logger.debug("Can't uppercase something that isn't a string",
-                      :field => field, :value => event[field])
-      end
+      original = event[field]
+      event[field] = case original
+        when Array
+          original.map(&:upcase!)
```

Comment:
But that is exactly the bug! ```event[field]``` in the case of JavaProxy -> HashMap returns a copy.  I left the bang ! in because as it is a copy the not bang method will make a new String instance of the copy.  We have to assign the pointer to the modified copy (as supplied by JavaProxy) back the the event[field].  This is why the Event should deprecate ```[] and []=``` in favour of get and set thereby making it more explicit to the plugin author looking to modify an Event.